### DISCUSSION TITLE
Remove CS1717 warning after not null assertion

### DISCUSSION THREAD

**Discussion by neoaisac** [dotnet/roslyn#46583] [General] [answered]
**Version Used**: 

**Steps to Reproduce**:

1. With NUnit, assert a nullable variable to be not null.
2. Make further assertions after this.
3. The `CS1717` warning will appear on the first line after the not-null assertion.
4. This happens even if the suppression operator is used on the not-null assertion.

```
Assert.NotNull(result!); // Suppression operator should suppress CS1717

Assert.Equals("value", result.Property); // CS1717 warning still appears here
Assert.Equals("value", result.OtherProperty); // CS1717 warning will not appear here
```

If the developer chooses to use the suppression (bang) operator on the first line after the first assertion, the CS1717 warning will be removed, but if future refactoring causes reordering of the lines, the operator might have to be placed on the first line again, or, annoyingly, on all lines, to ensure that if lines are reordered they don't start getting the warning.

**Expected Behavior**:

Ideally, Roslyn should understand that a not null assertion is equivalent to throwing an exception if null. Else, it should understand that the suppression operator added to such a line should affect all lines below.

**Actual Behavior**:

Roslyn does not understand that the `Assert.NotNull` assertion should have the same effect as a nullability check, nor does it understand the suppression operator added to the variable in the `NotNull` assertion parameter.

**Comment by RikkiGibson** [comment, marked as answer]
Re: a `!` suppression operator updating the operand state to not-null. I assume the C# language team has considered this question at length and is unlikely to change the behavior. /cc @jcouv who may be able to provide context on that specifically.

It looks like Nunit is unable to put `[NotNull]` on the parameter of their `Assert.NotNull` method because the method doesn't throw on a null argument in the context of a lambda passed to `Assert.Multiple`. You may be able to solve your problem by introducing API within your own test project similar to the following:
```cs
using System.Diagnostics.CodeAnalysis;

void AssertNotNull([NotNull] object? arg)
{
    Assert.NotNull(arg);
    if (arg is null)
    {
        throw new ArgumentNullException(nameof(arg), "Unexpected call from Assert.Multiple context");
    }
}
```

For some time Roslyn introduced `RoslynDebug.Assert` and similar helpers to hint information like this to flow analysis.

**Comment by RikkiGibson** [comment]
Also, CS1717 is for "assignment to self", is that actually the warning that appears in the code above? That seems strange.
